File: frontend/src/app/models/investigation.ts
/**
 * TypeScript models for Solar Investigation management
 * These models match the backend Pydantic models for API compatibility
 */

export enum InvestigationStatus {
    PENDING = 'pending',
    RUNNING = 'running',
    COMPLETED = 'completed',
    FAILED = 'failed',
    CANCELLED = 'cancelled'
}

export enum AgentMessageType {
    USER = 'user',
    AGENT = 'agent',
    SYSTEM = 'system',
    THINKING = 'thinking',
    TOOL_CALL = 'tool_call',
    TOOL_RESULT = 'tool_result'
}

export interface AgentMessage {
    id: string;
    investigation_id: string;
    message_type: AgentMessageType;
    content: string;
    metadata?: { [key: string]: any };
    timestamp: string; // ISO string from backend

    // UI State fields
    ui_summary?: string; // 10-word UI summary for display
    ui_state?: { [key: string]: any }; // UI-specific state data
    show_full_content?: boolean; // Whether to show full content by default
}

export interface InvestigationRequest {
    plant_id: string;
    start_date: string; // ISO date string (YYYY-MM-DD)
    end_date: string; // ISO date string (YYYY-MM-DD)
    additional_notes?: string;
    parent_id?: string; // ID of parent investigation if this is a retry
}

export interface Investigation {
    id: string;
    plant_id: string;
    plant_name?: string; // Will be populated from plant data
    start_date: string; // ISO date string
    end_date: string; // ISO date string
    additional_notes?: string;
    status: InvestigationStatus;
    session_id?: string;
    user_id: string;
    parent_id?: string; // ID of parent investigation if this is a retry
    created_at: string; // ISO string from backend
    updated_at: string; // ISO string from backend
    completed_at?: string;
    result?: { [key: string]: any };
    error_message?: string;
    agent_stats?: { // Rich ADK agent interaction statistics
        total_events: number;
        user_messages: number;
        agent_responses: number;
        thinking_steps: number;
        tool_calls: number;
        tools_used: string[];
        total_agents: string[];
        session_duration?: number;
        last_activity?: string;
        progress_steps: Array<{
            step_number: number;
            step_name: string;
            timestamp: string;
            completed: boolean;
        }>;
    };
}

export interface InvestigationResponse {
    investigation: Investigation;
    message: string;
}

export interface InvestigationListResponse {
    investigations: Investigation[];
    total: number;
    page: number;
    size: number;
}

export interface ChatHistoryResponse {
    investigation_id: string;
    messages: AgentMessage[];
    total_messages: number;
}

export interface DecisionRequest {
    decision: string;
    decision_type?: string;
    notes?: string;
}

export interface DecisionResponse {
    success: boolean;
    message: string;
    updated_investigation: Investigation;
}

export interface MessageRequest {
    content: string;
    message_type: AgentMessageType;
}

export interface HealthResponse {
    status: string;
    version: string;
    timestamp: string;
    database: {
        status: string;
        total_investigations: number;
    };
}

// Workorder related interfaces
export enum WorkorderStatus {
    PENDING = 'pending',
    IN_PROGRESS = 'in_progress',
    COMPLETED = 'completed',
    FAILED = 'failed'
}

export enum WorkorderType {
    MAINTENANCE = 'maintenance',
    INSPECTION = 'inspection',
    REPAIR = 'repair',
    ANALYSIS = 'analysis'
}

export interface WorkorderAgentRequest {
    todo_summary: string;
    priority?: string;
}

export interface WorkorderAgentResponse {
    workorder_id: string;
    investigation_id: string;
    todo_summary: string;
    agent_response: string;
    priority: string;
    status: WorkorderStatus;
    workorder_type: WorkorderType;
    created_at: string;
}

export interface Workorder {
    id: string;
    investigation_id: string;
    type: WorkorderType;
    description: string;
    agent_response?: string;
    priority: string;
    status: WorkorderStatus;
    created_at: string;
    updated_at: string;
    completed_at?: string;
}

// Legacy interface for backward compatibility with existing dashboard component
export interface LegacyInvestigation {
    id: string;
    title: string;
    status: 'running' | 'complete' | 'failed';
    agentChat: LegacyAgentMessage[];
    findings?: string;
    createdAt: Date;
    projectId?: string;
}

export interface LegacyAgentMessage {
    agentName: string;
    message: string;
    timestamp: Date;
}

File: backend/src/adk/models/investigation.py
"""Data models for Solar Investigation management"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, date
from enum import Enum
import uuid


class InvestigationStatus(str, Enum):
    """Investigation status enumeration"""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    REQUIRES_ATTENTION = "requires_attention"
    FAILED = "failed"
    CANCELLED = "cancelled"


class AgentMessageType(str, Enum):
    """Agent message type enumeration"""

    USER = "user"
    AGENT = "agent"
    SYSTEM = "system"
    THINKING = "thinking"
    TOOL_CALL = "tool_call"
    TOOL_RESULT = "tool_result"


class AgentMessage(BaseModel):
    """Agent message model for chat history with UI state support"""

    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    investigation_id: str
    message_type: AgentMessageType
    content: str  # Full content
    metadata: Optional[Dict[str, Any]] = None
    timestamp: datetime = Field(default_factory=datetime.now)

    # UI State fields for frontend display
    ui_summary: Optional[str] = Field(
        None, description="10-word UI summary for display"
    )
    ui_state: Optional[Dict[str, Any]] = Field(
        None, description="UI-specific state data"
    )
    show_full_content: bool = Field(
        False, description="Whether to show full content by default"
    )

    class Config:
        json_encoders = {datetime: lambda v: v.isoformat()}


class InvestigationRequest(BaseModel):
    """Request model for starting new investigation"""

    plant_id: str = Field(..., description="ID of the plant to investigate")
    start_date: date = Field(..., description="Start date for investigation period")
    end_date: date = Field(..., description="End date for investigation period")
    additional_notes: Optional[str] = Field(
        None, description="Additional notes or requirements"
    )
    parent_id: Optional[str] = Field(
        None, description="ID of parent investigation if this is a retry"
    )


class Investigation(BaseModel):
    """Investigation data model"""

    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    plant_id: str
    start_date: date
    end_date: date
    additional_notes: Optional[str] = None
    status: InvestigationStatus = InvestigationStatus.PENDING
    session_id: Optional[str] = None
    user_id: str = "api_user"
    parent_id: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    completed_at: Optional[datetime] = None
    result: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    agent_stats: Optional[Dict[str, Any]] = None  # Rich ADK agent interaction stats

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat(),
            date: lambda v: v.isoformat(),
        }


class InvestigationResponse(BaseModel):
    """Response model for investigation endpoints"""

    investigation: Investigation
    message: str


class InvestigationListResponse(BaseModel):
    """Response model for investigation list"""

    investigations: List[Investigation]
    total: int
    page: int
    size: int


class ChatHistoryResponse(BaseModel):
    """Response model for chat history"""

    investigation_id: str
    messages: List[AgentMessage]
    total_messages: int


class DecisionRequest(BaseModel):
    """Request model for human decisions during investigation"""

    decision: str = Field(..., description="Human decision or response")
    decision_type: str = Field(
        default="continue", description="Type of decision (continue, stop, modify)"
    )
    additional_data: Optional[Dict[str, Any]] = None


class DecisionResponse(BaseModel):
    """Response model for decision handling"""

    investigation_id: str
    decision_accepted: bool
    message: str
    next_steps: Optional[List[str]] = None
